collector/collector.go

```diff
@@ -3,161 +3,109 @@ package collector
-	}
-
-	err = parseRequestBody(resp, ret)
-	if err != nil {
-		return fmt.Errorf("failed to process response body: %w", err)
+func (c LCDCollector) Get(metric Metric) (float64, error) {
+	monitor, found := c.Metrics[metric]
+	if !found {
+		return 0, fmt.Errorf("monitor for metric \"%s\" not found", metric)
 	}
-	return nil
+	return monitor.Get(metric)
 }
 
-func (c LCDCollector) buildAndProcessRequest(
-	ctx context.Context,
-	contractAddress string,
-	query interface{},
-	ret interface{},
-) error {
-	req, err := c.buildRequest(ctx, contractAddress, query)
-	if err != nil {
-		return fmt.Errorf(
-			"failed to build request %+v for %s contract: %w",
-			query,
-			contractAddress,
-			err,
-		)
-	}
-	err = c.processRequest(req, ret)
-	if err != nil {
-		return fmt.Errorf("failed to process %+v request for %s contract: %w", query, contractAddress, err)
+func (c *LCDCollector) UpdateData(ctx context.Context) error {
+	for _, monitor := range c.Monitors {
+		err := monitor.Handler(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to update data: %w", err)
+		}
 	}
 	return nil
 }
 
-func (c *LCDCollector) getRewardState(ctx context.Context) (*RewardStateResponse, error) {
-	req, resp := GetRewardResponseStatePair()
-	err := c.buildAndProcessRequest(ctx, c.RewardContractAddress, req, &resp)
-	if err != nil {
-		return nil, fmt.Errorf("failed to process RewardState request: %w", err)
-	}
-	c.logger.Infoln("updated reward state info")
-	return &resp, nil
-}
-
-func (c *LCDCollector) getBlunaTokenInfo(ctx context.Context) (*TokenInfoResponse, error) {
-	req, resp := GetCommonTokenInfoPair()
-	err := c.buildAndProcessRequest(ctx, c.BlunaContractAddress, req, &resp)
-	if err != nil {
-		return nil, fmt.Errorf("failed to process BlunaTokenInfo request: %w", err)
+func (c *LCDCollector) RegisterMonitor(m Monitor) {
+	for _, metric := range m.ProvidedMetrics() {
+		if wantedMonitor, found := c.Metrics[metric]; found {
+			panic(fmt.Sprintf("register monitor %s failed. metrics collision. Monitor %s has declared metric %s", m.Name(), wantedMonitor.Name(), metric))
+		}
+		c.Metrics[metric] = m
+		_, err := m.Get(metric)
+		var doesNotExistError *MetricDoesNotExistError
+		if err != nil && errors.As(err, &doesNotExistError) {
+			panic(fmt.Sprintf("register monitor %s failed. Metric validation error. %+v", m.Name(), err))
+		}
 	}
-	c.logger.Infoln("updated bluna token info")
-	return &resp, nil
+	c.Monitors = append(c.Monitors, m)
+	m.SetApiClient(c.apiClient)
+	m.SetLogger(c.logger)
 }
 
-func (c *LCDCollector) UpdateData(ctx context.Context) error {
-	blunaTokenInfo, err := c.getBlunaTokenInfo(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to get bluna token info: %w", err)
-	}
-	c.CollectedData.BlunaTokenInfo = *blunaTokenInfo
-
-	rewardState, err := c.getRewardState(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to get reward state: %w", err)
-	}
-	c.CollectedData.RewardState = *rewardState
-	return nil
+type Monitor interface {
```

Comment:
Maybe we should move the Monitor type (along with monitor implementations) to a separate package?

Reply:
monitors moved to terra-monitors/collector/monitors package